devito/yask/data.py

```diff
@@ -0,0 +1,311 @@
+                    # the padding size will be automatically increase to accomodate it."
+                    self.grid.set_halo_size(i.name, halo)
+
+            # Allocate memory
+            self.grid.alloc_storage()
+
+            self._halo = []
+            self._ofs = [0 if i.is_Time else self.get_first_rank_domain_index(i.name)
+                         for i in dimensions]
+            self._shape = shape
+
+            # `self` will actually act as a view of `self.base`
+            self.base = Data(grid, shape, dimensions, 0, dtype)
+
+            # Initialize memory to 0
+            self.reset()
+        else:
+            self._halo = [0 if i.is_Time else self.get_halo_size(i.name)
+                          for i in dimensions]
+            self._ofs = [0 if i.is_Time else (self.get_first_rank_domain_index(i.name)-j)
+                         for i, j in zip(dimensions, self._halo)]
+            self._shape = [i + 2*j for i, j in zip(shape, self._halo)]
+
+            # Like numpy.ndarray, `base = None` indicates that this is the real
+            # array, i.e., it's not a view
+            self.base = None
+
+    def __getitem__(self, index):
+        start, stop, shape = self._convert_index(index)
+        if not shape:
+            log("Data: Getting single entry %s" % str(start))
+            assert start == stop
+            out = self.grid.get_element(start)
+        else:
+            log("Data: Getting full-array/block via index [%s]" % str(index))
+            out = np.empty(shape, self.dtype, 'C')
+            self.grid.get_elements_in_slice(out.data, start, stop)
+        return out
+
+    def __setitem__(self, index, val):
+        start, stop, shape = self._convert_index(index, 'set')
+        if all(i == 1 for i in shape):
+            log("Data: Setting single entry %s" % str(start))
+            assert start == stop
+            self.grid.set_element(val, start)
+        elif isinstance(val, np.ndarray):
+            log("Data: Setting full-array/block via index [%s]" % str(index))
+            self.grid.set_elements_in_slice(val, start, stop)
+        elif all(i == j-1 for i, j in zip(shape, self.shape)):
+            log("Data: Setting full-array to given scalar via single grid sweep")
+            self.grid.set_all_elements_same(val)
+        else:
+            log("Data: Setting block to given scalar via index [%s]" % str(index))
+            self.grid.set_elements_in_slice_same(val, start, stop, True)
+
+    def __getslice__(self, start, stop):
+        if stop == sys.maxint:
+            # Emulate default NumPy behaviour
+            stop = None
+        return self.__getitem__(slice(start, stop))
+
+    def __setslice__(self, start, stop, val):
+        if stop == sys.maxint:
+            # Emulate default NumPy behaviour
+            stop = None
+        self.__setitem__(slice(start, stop), val)
+
+    def _convert_index(self, index, mode='get'):
```

Comment:
Note for reviewers: this was formerly a free function in ``devito.yask.utils.py``, called ``convert_multislice``